### PR TITLE
Add long running activity tag for SignalR connections

### LINKED ISSUE #29699: Add information to indicate a SignalR request is long running
To allow APMs (e.g., Application Insights) to treat long running SignalR requests differently from the their request statistics, we should set some information (I presume via an `Activity.Tag`?) on current Activity to indicate a request is long running. 

Edit: After more discussion with App Insights, we think headers will work.

cc @cijothomas

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/29699

### CODE REVIEW

File: src/SignalR/common/Http.Connections/test/LongPollingTests.cs [hunk 1]
```diff
@@ -116,5 +116,25 @@ public void CheckLongPollingTimeoutValue()
...
             Assert.Equal(options.LongPolling.PollTimeout, TimeSpan.FromSeconds(90));
         }
+
+        [Fact]
+        public async Task IncludesLongRunningHeader()
+        {
+            using (StartVerifiableLog())
+            {
+                var pair = DuplexPipe.CreateConnectionPair(PipeOptions.Default, PipeOptions.Default);
+                var connection = new DefaultConnectionContext("foo", pair.Transport, pair.Application);
```
Comment (bradygaster) [inline src/SignalR/common/Http.Connections/test/LongPollingTests.cs:126]: I thought I read somewhere `foo` was never good in public-facing code/tests but nbd.

File: src/SignalR/common/Http.Connections/test/ServerSentEventsTests.cs [hunk 2]
```diff
@@ -131,6 +131,25 @@ public async Task SSEAddsAppropriateFraming(string message, string expected)
             }
         }
 
+        [Fact]
+        public async Task IncludesLongRunningHeader()
+        {
+            using (StartVerifiableLog())
+            {
+                var pair = DuplexPipe.CreateConnectionPair(PipeOptions.Default, PipeOptions.Default);
+                var connection = new DefaultConnectionContext("foo", pair.Transport, pair.Application);
```
Comment (bradygaster) [inline src/SignalR/common/Http.Connections/test/ServerSentEventsTests.cs:140]: and also but again nbd just making sure if you change one you change both. :)

File: src/SignalR/common/Http.Connections/test/WebSocketsTests.cs [hunk 3]
```diff
@@ -418,5 +418,48 @@ public async Task MultiSegmentSendWillNotSendEmptyEndOfMessageFrame()
...
             }
         }
+
+        [Fact]
+        public async Task IncludesLongRunningHeader()
+        {
+            using (StartVerifiableLog())
+            {
+                var pair = DuplexPipe.CreateConnectionPair(PipeOptions.Default, PipeOptions.Default);
+                var connection = new HttpConnectionContext("foo", connectionToken: null, LoggerFactory.CreateLogger(nameof(HttpConnectionContext)), pair.Transport, pair.Application);
```
Comment (bradygaster) [inline src/SignalR/common/Http.Connections/test/WebSocketsTests.cs:428]: all three pardon

**Review by bradygaster — APPROVED** [review]

File: src/SignalR/common/Http.Connections/test/LongPollingTests.cs (hunk 1 above)
Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/test/LongPollingTests.cs:126]: fubar isn't allowed generally. But I think foo or bar separately is fine.

Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/test/LongPollingTests.cs:126]: fubar isn't allowed generally. But I think foo or bar separately is fine.

File: src/SignalR/common/Http.Connections/src/Internal/Transports/CustomHeaderNames.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Http.Connections.Internal.Transports
+{
+    internal static class CustomHeaderNames
+    {
+        /// <summary>
+        /// Used as a response header to let other parties, like telemetry systems, know that the request/connection will take a long time.
+        /// </summary>
+        public static readonly string LongRunning = "Long-Running";
```
Comment (halter73) [inline src/SignalR/common/Http.Connections/src/Internal/Transports/CustomHeaderNames.cs:11]: It'd be nice if there was prior art for this, but I'm glad we're [not using X-](https://tools.ietf.org/html/rfc6648).

**Review by halter73 — APPROVED** [review]

**Comment by davidfowl** [comment]
Should this be moved to the websocket middleware instead of signalr?

**Comment by shirhatti** [comment]
I thought we we're doing to the header in addition to setting a tag on the activity?

**Comment by shirhatti** [comment]
@cijothomas I assume this will be sufficient for AI to exclude long-running SigalR requests from regular request statistics?

**Comment by davidfowl** [comment]
> I thought we we're doing to the header in addition to setting a tag on the activity?

Why?

**Comment by rajkumar-rangaraj** [comment]
> @cijothomas I assume this will be sufficient for AI to exclude long-running SigalR requests from regular request statistics?

@shirhatti This will help in AI to exclude long-running requests of Signal-R.

@BrennanConroy

* Application insights gets notified on new ASP.NET Core / SignalR request when it enters `Microsoft.AspNetCore.Hosting.HostingApplicationDiagnostics.BeginRequest`. At this point we create a request telemetry object to track the request.
* When `Microsoft.AspNetCore.Hosting.HostingApplicationDiagnostics.RequestEnd` is invoked AI gets notified again at this point we update request telemetry with duration and send it to application insights service.

With the current change we have to create and hold the request telemetry in the heap till the `RequestEnd` is called and drop request telemetry object on the floor. Proposed change will help us suppress the long running requests but there is a perf issue in creating and holding the RequestTelemetry. Is is possible to add HttpContext request header in BeginRequest before start activity is called.

**Comment by shirhatti** [comment]
>  Is it possible to add HttpContext request header in BeginRequest before start activity is called.

That's impossible. StartActivity is called way too early for that

**Comment by shirhatti** [comment]
@davidfowl So that I don't have to use DiagnosticsListener, get the HttpContext and then read the response headers. If I'm an instrumentation library I just want to have all the info on the Activity

**Comment by shirhatti** [comment]
Also @rajkumar-rangaraj ,we're adding a response header, not a request header 😄

**Comment by rajkumar-rangaraj** [comment]
> Also @rajkumar-rangaraj ,we're adding a response header, not a request header 😄

Yes. I reviewed it. As I wanted an update in BeginRequest asked for header in request header.

**Comment by davidfowl** [comment]
> Yes. I reviewed it. As I wanted an update in BeginRequest asked for change in request header.

This isn't possible.

**Comment by halter73** [comment]
> Should this be moved to the websocket middleware instead of signalr's websocket transport.

I thought you might want to make the WebSocket middleware a little more raw. Instrumentation should treat all WebSocket requests as long-running to begin with.

**Comment by BrennanConroy** [comment]
Updated

**Review by bradygaster — APPROVED** [review]

**Comment by davidfowl** [comment]
nit: Fix the PR title 😄

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs [hunk 5]
```diff
@@ -115,6 +115,7 @@ public async Task ExecuteNegotiateAsync(HttpContext context, HttpConnectionDispa
 
         private async Task ExecuteAsync(HttpContext context, ConnectionDelegate connectionDelegate, HttpConnectionDispatcherOptions options, ConnectionLogScope logScope)
         {
+            context.Features.Get<IHttpActivityFeature>()?.Activity.AddTag("http.long_running", "true");
```
Comment (davidfowl) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs:118]: Add a comment.

**Review by shirhatti — APPROVED** [review]

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs [hunk 6]
```diff
@@ -115,6 +115,9 @@ public async Task ExecuteNegotiateAsync(HttpContext context, HttpConnectionDispa
 
         private async Task ExecuteAsync(HttpContext context, ConnectionDelegate connectionDelegate, HttpConnectionDispatcherOptions options, ConnectionLogScope logScope)
         {
+            // set a tag to allow Application Performance Management tools to differentiate long running requests for reporting purposes
```
Comment (davidfowl) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs:118]: I haven't seen APM written out like this before

**Review by davidfowl — APPROVED** [review]

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs (hunk 6 above)
Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs:118]: I didn't know what it stood for and in 1 year I definitely wont remember.

**Pull request merged**